---
title: "Simplifying Imaging-based Spatial Transcriptomics Data Analysis"
date: "`r BiocStyle::doc_date()`"
author:
- name: BÃ¡rbara Zita Peters Couto
  affiliation:  
  - "???"
vignette: >
  %\VignetteIndexEntry{"Introduction to simpleSeg"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(BiocStyle)
```

# MoleculeExperiment
The R package MoleculeExperiment contains functions to create and work with
objects from the new MoleculeExperiment class. We introduce this class for
analysing molecule-based spatial transcriptomics data (e.g., Xenium by 10X,
Cosmx SMI by Nanostring, and Merscope by Vizgen).

## Why the MoleculeExperiment class?
The goal of the MoleculeExperiment class is to:
1) Enable analysis of spatial transcriptomics data at the molecule level, 
rather than the cell level like other S4 classes currently do.
2) Standardise molecule-based ST data across vendors, to hopefully facilitate
comparison of different data sources.

# Minimal example
First load required libraries.

```{r}
# load necessary libraries
library(MoleculeExperiment)
library(ggplot2)

```

To create an ME object (i.e., object of the MoleculeExperiment class), first use
the constructor.

```{r}
# TODO make ME
# USE LITTLE ARGUMENTS HERE
# use toy example data here
#(but do NOT print the output, the focus here is on the quick code)
# only printe output of how the ME looks like

```

The MoleculeExperiment class stores data in two slots: molecules and boundaries.
Both slots have the same hierarchical nested list structure.

One can then use this data to do molecule-level analyses as in the figure below.

```{r}
# visualise ME
# change code below to work with molecules(toy_me, "detected", "flatten = TRUE))

#ggplot(molecules_df, aes(x = x_location, y = y_location)) +
#  geom_point() +
#  geom_polygon(aes(group = cell_id),
#                    fill = NA,
#                    colour = "black",
#                    data = boundaries_df) +
#  facet_wrap(~sample_id) +
#  coord_fixed()

```

Finally, it is also possible to go from a MoleculeExperiment object to a
SpatialExperiment object. This enables the transition from a molecule-level
analysis to a cell-level analysis with already existing tools.

```{r}
# transform ME to SPE object
spe <- countMolecules(me = toy_me)
```

# The ME object in detail

## Constructing an ME object

### Use case 1: from dataframes to ME object 
Here we demonstrate how to generate an ME object with toy data, representing a 
scenario where both the detected transcripts information and the boundary 
information have already been read into R. 

First we generate a toy transcripts data.frame:

```{r}
# molecules data.frame toy example
molecules_df <- data.frame(
  sample_id = rep(c("sample1", "sample2"), times = c(30, 20)),
  feature_name = rep(c("gene1", "gene2"), times = c(20, 30)),
  x_location = runif(50),
  y_location = runif(50)
  )
head(molecules_df)
```

Then we generate a toy boundaries data.frame:

```{r}
# boundaries data.frame toy example
boundaries_df <- data.frame(
  sample_id = rep(c("sample1", "sample2"), times = c(16, 6)),
  cell_id = rep(c("cell1", "cell2", "cell3", "cell4",
                  "cell1", "cell2"),
                times = c(4, 4, 4, 4, 3, 3)),
  x_location = c(0, 0.5, 0.5, 0,
                 0.5, 1, 1, 0.5,
                 0, 0.5, 0.5, 0,
                 0.5, 1, 1, 0.5,
                 0, 1, 0,
                 0, 1, 1),
  y_location = c(0, 0, 0.5, 0.5,
                 0, 0, 0.5, 0.5,
                 0.5, 0.5, 1, 1,
                 0.5, 0.5, 1, 1,
                 0, 1, 1,
                 0, 0, 1)
)
head(boundaries_df)
```

To generate an ME object, the next step is to standardise this data to the 
MoleculeExperiment list format.


```{r}
molecules_ls <- dataframeToMEList(molecules_df,
                                  df_type = "transcripts",
                                  assay_name = "detected",
                                  sample_col = "sample_id",
                                  factor_col = "feature_name",
                                  x_col = "x_location",
                                  y_col = "y_location")
# to avoid printing large nested list in the terminal, use str() and the
# max.level argument
str(molecules_ls, max.level = 3)
```


```{r}
boundaries_ls <- dataframeToMEList(boundaries_df,
                                   df_type = "boundaries",
                                   assay_name = "cells",
                                   sample_col = "sample_id",
                                   factor_col = "cell_id",
                                   x_col = "x_location",
                                   y_col = "y_location")
str(boundaries_ls, 3)
```

Now that the transcript and boundary information is in a standardised ME list
format, it can be used as input to generate an ME object.

```{r}
# use MoleculeExperiment object constructor
toy_me <- MoleculeExperiment(molecules = molecules_ls,
                          boundaries = boundaries_ls)

# visualise contents of me object
toy_me
```

### Use case 2: from Xenium to ME object
Due to the flexibility of the function arguments, the above use case is useful
for creating an ME object from dataframes of various different vendors. Nevertheless,
the MoleculeExperiment package also provides functions to directly work with
the directories containing output data of commonly used technologies. This is
especially useful to work with data from multiple samples.

Here we show how one can use functions from the MoleculeExperiment package to
read in Xenium data (10X Genomics) into an ME object.

```{r}
# create ME obj with example mouse brain xenium dataset
repo_dir <- "/dski/nobackup/bpeters/SpatialUtils/inst/extdata/mouse_brain_mini_xenium"
me <- readXenium(repo_dir,
                  n_samples = 2,
                  keep_cols = "essential",
                  add_boundaries = c("cell", "nucleus"))
# visualise me contents
me
```

In Xenium, the transcript.csv file has x and y locations as microns, whereas
segmentation files have their locations in pixels. Our function readXenium takes
this into account, and standardises the column contents to the micron units. 

readXenium calls readMolecules and readBoundaries under the hood. These 
convenience function can also be used by themselves, which might be useful in
this context where new molecule-level spatial transcriptomics technologies are
being developed.

readMolecules enables the user to decide if they want to keep all the 
data that is vendor-specific (e.g., qv in xenium).


## ME object structure
A MoleculeExperiment object contains a @molecules slot and a @boundaries slot.

Both slots have a hierarchical list structure that consists of a nested list,
ultimately ending in a data.frame/tibble. This data structure was used to work
with an object of less size. Traditional rectangular data structures, like
dataframes, redundantly store gene names and sample IDs for the millions of
transcripts. In contrast, data in a list enables us to avoid this redundancy and
work with objects of smaller size.


### molecules slot
The @molecules slot contains molecule-level information. The essential data it 
contains is the feature name (e.g., gene names) and x and y locations of the
detected molecules (e.g., transcripts), in each sample. Nevertheless, the user
can also decide to keep all molecule metadata (e.g., subcellular location:
nucleus/cytoplasm).

The nested list in the molecules slot has the following hierarchical structure:
"assay name" > "sample ID" > "feature name" > dataframe/tibble with X and Y
locations (and other additional columns of interest).

```{r}
strMolecules(me)
```

### boundaries slot
The @boundaries slot contains information from segmentation analyses (e.g.,
cell boundaries, or nucleus boundaries).

The nested list in the boundaries slot has the following hierarchical structure:
"assay name" > "sample ID" > "segment ID" > dataframe/tibble with the vertex
coordinates defining the boundaries for each segment. For example, if the
boundary information is for cells, the assay name can be set to "cell";
or "nucleus" if one is using nucleus boundaries.

```{r}
strBoundaries(me)
```

# Methods
Here we introduce basic methods to access (getters) and manipulate (setters)
data in an ME object.

## Getters
The main getters are molecules() and boundaries().

```{r}
# note that output from the following methods can be very large.
# These getters should be used when the data from the slots needs to be used as
# input for other functions.

# molecules(me) or molecules(me, "detected")
identical(molecules(me), molecules(me, "detected"))

molecules(me)[[1]][[1]][[1]]

# it is recommended to use strMolecules or strBoundaries when trying to
# get a quick visualisation of the data.
```

```{r}
# boundaries(me, "cells")
boundaries(me, "cells")[[1]][[1]][[1]]
```

Other getters include: features() and segmentIDs().

```{r}
features(me)
```

```{r}
segmentIDs(me, "cells")
```

## Setters
Main setters include molecules<- and boundaries<-.
For example, with boundaries<- one can add new segmentation assay information
to the boundaries slot. 

```{r}
nuclei_ls <- readBoundaries(data_dir = repo_dir,
                            pattern = "nucleus_boundaries.csv",
                            n_samples = 2,
                            segment_id_col = "cell_id",
                            x_col = "vertex_x",
                            y_col = "vertex_y",
                            keep_cols = "essential",
                            boundaries_assay = "nucleus")

# add nucleus boundaries to already existing ME object
boundaries(me, "nucleus") <- nuclei_ls
me # note the addition of the nucleus boundaries to the nucleus slot
# note that one can also assign new lists to the molecules slot with molecules<-
```


# From MoleculeExperiment to SpatialExperiment
The idea behind the MoleculeExperiment class is to 1) store data from
molecule-based spatial transcriptomics technologies in a way that enables a
molecule-level analysis, and 2) standardise output data format across the many
different vendors.
If one is interested in continuing downstream analysis at the cell-level,
the MoleculeExperiment package also provides a convenience function,
countMolecules(), that enables the transition from a MoleculeExperiment
object to a SpatialExperiment object. With this functionality, it is possible to
use already existing methods developed to analyse cell-level data stored as a
SpatialExperiment object.

```{r}
spe <- countMolecules(me = me)
spe_nuc <- countMolecules(me = me, boundaries_assay = "nucleus")

```

```{r}
# quantify counts associated to cells versus nuclei
colData(spe)
SummarizedExperiment::assay(spe, "nucleus") <- SummarizedExperiment::assay(spe_nuc, "counts")
```

```{r}
plot(as.numeric(SummarizedExperiment::assay(spe, "counts")),
  as.numeric(SummarizedExperiment::assay(spe, "nucleus")))
```


# Future developments
The MoleculeExperiment package introduces the MoleculeExperiment class. In the
future, a new package called SpatialUtils will introduce functional programming
capabilities to work with the MoleculeExperiment object.

## Session Info
```{r}
sessionInfo()
```
