<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MoleculeExperiment">
<title>An introduction to the MoleculeExperiment Class • MoleculeExperiment</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the MoleculeExperiment Class">
<meta property="og:description" content="MoleculeExperiment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MoleculeExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/MoleculeExperiment.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/MoleculeExperiment/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="MoleculeExperiment_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>An introduction to the MoleculeExperiment Class</h1>
                        <h4 data-toc-skip class="author">Bárbara Zita Peters Couto</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney, Sydney, NSW, 2006, AustraliaCharles Perkins Centre, The University of Sydney, Sydney, NSW, 2006, AustraliaSydney Precision Data Science Centre, The University of Sydney, Sydney, NSW, 2006, Australia<br><h4 data-toc-skip class="author">Nick Robertson</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney, Sydney, NSW, 2006, AustraliaCharles Perkins Centre, The University of Sydney, Sydney, NSW, 2006, AustraliaSydney Precision Data Science Centre, The University of Sydney, Sydney, NSW, 2006, Australia<br><h4 data-toc-skip class="author">Ellis Patrick</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney, Sydney, NSW, 2006, AustraliaCharles Perkins Centre, The University of Sydney, Sydney, NSW, 2006, AustraliaSydney Precision Data Science Centre, The University of Sydney, Sydney, NSW, 2006, AustraliaWestmead Institute for Medical Research, University of Sydney, Australia<br><h4 data-toc-skip class="author">Shila Ghazanfar</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney, Sydney, NSW, 2006, AustraliaCharles Perkins Centre, The University of Sydney, Sydney, NSW, 2006, AustraliaSydney Precision Data Science Centre, The University of Sydney, Sydney, NSW, 2006, Australia<br><h4 data-toc-skip class="date">12 April 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/MoleculeExperiment/blob/HEAD/vignettes/MoleculeExperiment.Rmd" class="external-link"><code>vignettes/MoleculeExperiment.Rmd</code></a></small>
      <div class="d-none name"><code>MoleculeExperiment.Rmd</code></div>
    </address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="moleculeexperiment">MoleculeExperiment<a class="anchor" aria-label="anchor" href="#moleculeexperiment"></a>
</h2>
<p>The R package MoleculeExperiment contains functions to create and work with objects from the new MoleculeExperiment class. We introduce this class for analysing molecule-based spatial transcriptomics data (e.g., Xenium by 10X, CosMx SMI by Nanostring, and Merscope by Vizgen, among others).</p>
<div class="section level3">
<h3 id="why-the-moleculeexperiment-class">Why the MoleculeExperiment class?<a class="anchor" aria-label="anchor" href="#why-the-moleculeexperiment-class"></a>
</h3>
<p>The goal of the MoleculeExperiment class is to: 1) Enable analysis of spatial transcriptomics (ST) data at the molecule level, independent of aggregating to the cell or tissue level. 2) Enable aggregation to a <code>SpatialExperiment</code> object given combinations of molecules and segmentation boundaries. 2) Standardise molecule-based ST data across vendors, to hopefully facilitate comparison of different data sources and common analytical and visualisation workflows.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The latest release of MoleculeExperiment can be installed using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"MoleculeExperiment"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="minimal-example">Minimal example<a class="anchor" aria-label="anchor" href="#minimal-example"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Load required libraries.</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/MoleculeExperiment" class="external-link">MoleculeExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Create MoleculeExperiment object with example Xenium data, taken over a small patch.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">repoDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"MoleculeExperiment"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">me</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readXenium.html">readXenium</a></span><span class="op">(</span><span class="va">repoDir</span>,</span>
<span>                  keepCols <span class="op">=</span> <span class="st">"essential"</span><span class="op">)</span></span>
<span><span class="va">me</span></span>
<span><span class="co">#&gt; class:  MoleculeExperiment </span></span>
<span><span class="co">#&gt; 2 samples: Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; @molecules contents: </span></span>
<span><span class="co">#&gt; -detected assay:</span></span>
<span><span class="co">#&gt; 178 unique features across all samples in assay "detected": 2010300C02Rik Acsbg1 Adamts2 Adamtsl1 Angpt1 Aqp4 ...</span></span>
<span><span class="co">#&gt; 869.5 molecules on average across all samples in assay "detected"</span></span>
<span><span class="co">#&gt; Location range across all samples in assay "detected": [4900,4919.98] x [6400.02,6420]</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Use standardised data in ME object for molecule-level analyses. For example, plot a simple digital in-situ, with cell boundaries overlaid.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="../reference/accessors.html">molecules</a></span><span class="op">(</span><span class="va">me</span>, flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_location</span>, y <span class="op">=</span> <span class="va">y_location</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_polygon.html" class="external-link">geom_polygon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">segment_id</span><span class="op">)</span>,</span>
<span>                    fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                    data <span class="op">=</span> <span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>,</span>
<span>                                      assayName <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                                      flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="MoleculeExperiment_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<ol start="4" style="list-style-type: decimal">
<li>Finally, it is also possible to go from a MoleculeExperiment object to a SpatialExperiment object. This enables the transition from a molecule-level analysis to a cell-level analysis with already existing tools.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># transform ME to SPE object</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countMolecules.html">countMolecules</a></span><span class="op">(</span><span class="va">me</span><span class="op">)</span></span>
<span><span class="va">spe</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 178 14 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(178): 2010300C02Rik Acsbg1 ... Trem2 Trp73</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(14): Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs.67500</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs.67512 ...</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs.65070</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs.65071</span></span>
<span><span class="co">#&gt; colData names(4): sample_id x_location y_location cell_id</span></span>
<span><span class="co">#&gt; reducedDimNames(1): spatial</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_location y_location</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-me-object-in-detail">The ME object in detail<a class="anchor" aria-label="anchor" href="#the-me-object-in-detail"></a>
</h2>
<div class="section level3">
<h3 id="constructing-an-me-object">Constructing an ME object<a class="anchor" aria-label="anchor" href="#constructing-an-me-object"></a>
</h3>
<div class="section level4">
<h4 id="use-case-1-from-dataframes-to-me-object">Use case 1: from dataframes to ME object<a class="anchor" aria-label="anchor" href="#use-case-1-from-dataframes-to-me-object"></a>
</h4>
<p>Here we demonstrate how to work with an ME object from toy data, representing a scenario where both the detected transcripts information and the boundary information have already been read into R. This requires the standardisation of the data with the dataframeToMEList() function.</p>
<p>The flexibility of the arguments in dataframeToMEList() enable the creation of a standard ME object across dataframes comming from different vendors of molecule-based spatial transcriptomics technologies.</p>
<ol style="list-style-type: decimal">
<li>Generate a toy transcripts data.frame.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moleculesDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  x_coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  y_coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">moleculesDf</span><span class="op">)</span></span>
<span><span class="co">#&gt;   sample_id features   x_coords  y_coords</span></span>
<span><span class="co">#&gt; 1   sample1    gene1 0.85495906 0.4467734</span></span>
<span><span class="co">#&gt; 2   sample1    gene1 0.99809022 0.9358170</span></span>
<span><span class="co">#&gt; 3   sample1    gene1 0.97953119 0.9806237</span></span>
<span><span class="co">#&gt; 4   sample1    gene1 0.24946555 0.7738943</span></span>
<span><span class="co">#&gt; 5   sample1    gene1 0.60894173 0.6703259</span></span>
<span><span class="co">#&gt; 6   sample1    gene1 0.06448276 0.3273922</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Generate a toy boundaries data.frame.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boundariesDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  cell_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1"</span>, <span class="st">"cell2"</span>, <span class="st">"cell3"</span>, <span class="st">"cell4"</span>,</span>
<span>                  <span class="st">"cell1"</span>, <span class="st">"cell2"</span><span class="op">)</span>,</span>
<span>                times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  vertex_x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0</span>, </span>
<span>               <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span>, </span>
<span>               <span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0</span>, </span>
<span>               <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span>,</span>
<span>               <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  vertex_y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, </span>
<span>               <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, </span>
<span>               <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>, </span>
<span>               <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>               <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">boundariesDf</span><span class="op">)</span></span>
<span><span class="co">#&gt;   sample_id cell_id vertex_x vertex_y</span></span>
<span><span class="co">#&gt; 1   sample1   cell1      0.0      0.0</span></span>
<span><span class="co">#&gt; 2   sample1   cell1      0.5      0.0</span></span>
<span><span class="co">#&gt; 3   sample1   cell1      0.5      0.5</span></span>
<span><span class="co">#&gt; 4   sample1   cell1      0.0      0.5</span></span>
<span><span class="co">#&gt; 5   sample1   cell2      0.5      0.0</span></span>
<span><span class="co">#&gt; 6   sample1   cell2      1.0      0.0</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Standardise transcripts dataframe to ME list format.</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moleculesMEList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataframeToMEList.html">dataframeToMEList</a></span><span class="op">(</span><span class="va">moleculesDf</span>,</span>
<span>                                  dfType <span class="op">=</span> <span class="st">"transcripts"</span>,</span>
<span>                                  assayName <span class="op">=</span> <span class="st">"detected"</span>,</span>
<span>                                  sampleCol <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>                                  factorCol <span class="op">=</span> <span class="st">"features"</span>,</span>
<span>                                  xCol <span class="op">=</span> <span class="st">"x_coords"</span>,</span>
<span>                                  yCol <span class="op">=</span> <span class="st">"y_coords"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">moleculesMEList</span>, max.level <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ detected:List of 2</span></span>
<span><span class="co">#&gt;   ..$ sample1:List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ gene1: tibble [20 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ gene2: tibble [10 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ sample2:List of 1</span></span>
<span><span class="co">#&gt;   .. ..$ gene2: tibble [20 × 2] (S3: tbl_df/tbl/data.frame)</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Standardise boundaries dataframe to ME list format.</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boundariesMEList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataframeToMEList.html">dataframeToMEList</a></span><span class="op">(</span><span class="va">boundariesDf</span>,</span>
<span>                                   dfType <span class="op">=</span> <span class="st">"boundaries"</span>,</span>
<span>                                   assayName <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                                   sampleCol <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>                                   factorCol <span class="op">=</span> <span class="st">"cell_id"</span>,</span>
<span>                                   xCol <span class="op">=</span> <span class="st">"vertex_x"</span>,</span>
<span>                                   yCol <span class="op">=</span> <span class="st">"vertex_y"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">boundariesMEList</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ cell:List of 2</span></span>
<span><span class="co">#&gt;   ..$ sample1:List of 4</span></span>
<span><span class="co">#&gt;   .. ..$ cell1: tibble [4 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ cell2: tibble [4 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ cell3: tibble [4 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ cell4: tibble [4 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ sample2:List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ cell1: tibble [3 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ cell2: tibble [3 × 2] (S3: tbl_df/tbl/data.frame)</span></span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>Create an ME object by using the MoleculeExperiment object constructor.</li>
</ol>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toyME</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MoleculeExperiment.html">MoleculeExperiment</a></span><span class="op">(</span>molecules <span class="op">=</span> <span class="va">moleculesMEList</span>,</span>
<span>                             boundaries <span class="op">=</span> <span class="va">boundariesMEList</span><span class="op">)</span></span>
<span><span class="va">toyME</span></span>
<span><span class="co">#&gt; class:  MoleculeExperiment </span></span>
<span><span class="co">#&gt; 2 samples: sample1 sample2 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; @molecules contents: </span></span>
<span><span class="co">#&gt; -detected assay:</span></span>
<span><span class="co">#&gt; 2 unique features across all samples in assay "detected": gene1 gene2 ...</span></span>
<span><span class="co">#&gt; 25 molecules on average across all samples in assay "detected"</span></span>
<span><span class="co">#&gt; Location range across all samples in assay "detected": [0.02,1] x [0.1,0.99]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="use-case-2-from-xenium-to-me-object">Use case 2: from Xenium to ME object<a class="anchor" aria-label="anchor" href="#use-case-2-from-xenium-to-me-object"></a>
</h4>
<p>The MoleculeExperiment package also provides functions to directly work with the directories containing output files of commonly used technologies. This is especially useful to work with data from multiple samples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">repoDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"MoleculeExperiment"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">me</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readXenium.html">readXenium</a></span><span class="op">(</span><span class="va">repoDir</span>,</span>
<span>                  keepCols <span class="op">=</span> <span class="st">"essential"</span>,</span>
<span>                  addBoundaries <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span>
<span><span class="va">me</span></span>
<span><span class="co">#&gt; class:  MoleculeExperiment </span></span>
<span><span class="co">#&gt; 2 samples: Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; @molecules contents: </span></span>
<span><span class="co">#&gt; -detected assay:</span></span>
<span><span class="co">#&gt; 178 unique features across all samples in assay "detected": 2010300C02Rik Acsbg1 Adamts2 Adamtsl1 Angpt1 Aqp4 ...</span></span>
<span><span class="co">#&gt; 869.5 molecules on average across all samples in assay "detected"</span></span>
<span><span class="co">#&gt; Location range across all samples in assay "detected": [4900,4919.98] x [6400.02,6420]</span></span></code></pre></div>
<p>readXenium() standardises the transcript and boundary information such that the column names are consistent across technologies when handling ME objects.</p>
<p>In addition, readXenium() enables the user to decide if they want to keep all data that is vendor-specific (e.g., column with qv score), some columns of interest, or only the essential columns. The latter refers to feature names and locations of the detected transcripts, and segment ids and boundary locations of the segmentation results.</p>
</div>
</div>
<div class="section level3">
<h3 id="me-object-structure">ME object structure<a class="anchor" aria-label="anchor" href="#me-object-structure"></a>
</h3>
<p>A MoleculeExperiment object contains a <span class="citation">@molecules</span> slot and a <span class="citation">@boundaries</span> slot.</p>
<p>Both slots have a hierarchical list structure that consists of a nested list, ultimately ending in a data.frame/tibble. Traditional rectangular data structures, like dataframes, redundantly store gene names and sample IDs for the millions of transcripts. In contrast, data in a list enables us to avoid this redundancy and work with objects of smaller size.</p>
<div class="section level4">
<h4 id="molecules-slot">molecules slot<a class="anchor" aria-label="anchor" href="#molecules-slot"></a>
</h4>
<p>The <span class="citation">@molecules</span> slot contains molecule-level information. The essential data it contains is the feature name (e.g., gene names) and x and y locations of the detected molecules (e.g., transcripts), in each sample. Nevertheless, the user can also decide to keep all molecule metadata (e.g., subcellular location: nucleus/cytoplasm).</p>
<p>The nested list in the molecules slot has the following hierarchical structure: “assay name” &gt; “sample ID” &gt; “feature name” &gt; dataframe/tibble with X and Y locations (and other additional columns of interest).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarization.html">strMolecules</a></span><span class="op">(</span><span class="va">me</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ detected:List of 2</span></span>
<span><span class="co">#&gt;   ..$ Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs:List of 137</span></span>
<span><span class="co">#&gt;   .. ..$ 2010300C02Rik        : tibble [11 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ Acsbg1               : tibble [6 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. .. [list output truncated]</span></span>
<span><span class="co">#&gt;   ..$ Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs:List of 143</span></span>
<span><span class="co">#&gt;   .. ..$ 2010300C02Rik: tibble [9 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ Acsbg1       : tibble [10 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. .. [list output truncated]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="boundaries-slot">boundaries slot<a class="anchor" aria-label="anchor" href="#boundaries-slot"></a>
</h4>
<p>The <span class="citation">@boundaries</span> slot contains information from segmentation analyses (e.g., cell boundaries, or nucleus boundaries).</p>
<p>The nested list in the boundaries slot has the following hierarchical structure: “assay name” &gt; “sample ID” &gt; “segment ID” &gt; dataframe/tibble with the vertex coordinates defining the boundaries for each segment. For example, if the boundary information is for cells, the assay name can be set to “cell”; or “nucleus” if one is using nucleus boundaries.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarization.html">strBoundaries</a></span><span class="op">(</span><span class="va">me</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ cell:List of 2</span></span>
<span><span class="co">#&gt;   ..$ Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs:List of 5</span></span>
<span><span class="co">#&gt;   .. ..$ 67500: tibble [13 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ 67512: tibble [13 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. .. [list output truncated]</span></span>
<span><span class="co">#&gt;   ..$ Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs:List of 9</span></span>
<span><span class="co">#&gt;   .. ..$ 65043: tibble [13 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ 65044: tibble [13 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. .. [list output truncated]</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>Here we introduce basic methods to access and manipulate data in an ME object, i.e., getters and setters, respectively.</p>
<div class="section level3">
<h3 id="getters">Getters<a class="anchor" aria-label="anchor" href="#getters"></a>
</h3>
<p>The main getters are molecules() and boundaries(). NOTE: the output of these methods is the ME nested list, which can be very large on screen. Thus, these getters should be used when wanting to work with the data. To quickly view the slot contents, use strMolecules() and strBoundaries() instead.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOTE: output not shown as it is too large</span></span>
<span><span class="co"># access molecules slot</span></span>
<span><span class="fu"><a href="../reference/accessors.html">molecules</a></span><span class="op">(</span><span class="va">me</span><span class="op">)</span></span>
<span><span class="co"># access cell boundary information in boundaries slot</span></span>
<span><span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>, <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
<p>For ease of use, these getters have arguments that enable the transformation of the data from a nested ME list format to a data.frame format.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/accessors.html">molecules</a></span><span class="op">(</span><span class="va">me</span>, assayName <span class="op">=</span> <span class="st">"detected"</span>, flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,739 × 4</span></span></span>
<span><span class="co">#&gt;    x_location y_location feature_name  sample_id                                </span></span>
<span><span class="co">#&gt;  <span style="color: #BCBCBC;">*</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>      <span style="text-decoration: underline;">4</span>918.      <span style="text-decoration: underline;">6</span>411. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>      <span style="text-decoration: underline;">4</span>901.      <span style="text-decoration: underline;">6</span>417. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>      <span style="text-decoration: underline;">4</span>901.      <span style="text-decoration: underline;">6</span>417. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>      <span style="text-decoration: underline;">4</span>910.      <span style="text-decoration: underline;">6</span>417. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>      <span style="text-decoration: underline;">4</span>908.      <span style="text-decoration: underline;">6</span>413. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>      <span style="text-decoration: underline;">4</span>911.      <span style="text-decoration: underline;">6</span>407. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>      <span style="text-decoration: underline;">4</span>915.      <span style="text-decoration: underline;">6</span>411. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>      <span style="text-decoration: underline;">4</span>916.      <span style="text-decoration: underline;">6</span>412. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>      <span style="text-decoration: underline;">4</span>901.      <span style="text-decoration: underline;">6</span>415. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>      <span style="text-decoration: underline;">4</span>906.      <span style="text-decoration: underline;">6</span>417. 2010300C02Rik Xenium_V1_FF_Mouse_Brain_MultiSection_1_…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,729 more rows</span></span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>, assayName <span class="op">=</span> <span class="st">"cell"</span>, flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 182 × 4</span></span></span>
<span><span class="co">#&gt;    x_location y_location segment_id sample_id                                   </span></span>
<span><span class="co">#&gt;  <span style="color: #BCBCBC;">*</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>      <span style="text-decoration: underline;">4</span>905.      <span style="text-decoration: underline;">6</span>400. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>      <span style="text-decoration: underline;">4</span>899.      <span style="text-decoration: underline;">6</span>401. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>      <span style="text-decoration: underline;">4</span>894.      <span style="text-decoration: underline;">6</span>408. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>      <span style="text-decoration: underline;">4</span>890.      <span style="text-decoration: underline;">6</span>418. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>      <span style="text-decoration: underline;">4</span>887.      <span style="text-decoration: underline;">6</span>423. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>      <span style="text-decoration: underline;">4</span>887.      <span style="text-decoration: underline;">6</span>425. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>      <span style="text-decoration: underline;">4</span>890.      <span style="text-decoration: underline;">6</span>427. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>      <span style="text-decoration: underline;">4</span>891.      <span style="text-decoration: underline;">6</span>427. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>      <span style="text-decoration: underline;">4</span>894.      <span style="text-decoration: underline;">6</span>426. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>      <span style="text-decoration: underline;">4</span>908.      <span style="text-decoration: underline;">6</span>421. 67500      Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 172 more rows</span></span></span></code></pre></div>
<p>Other getters include: features() and segmentIDs().</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get initial features in sample 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/accessors.html">features</a></span><span class="op">(</span><span class="va">me</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2010300C02Rik" "Acsbg1"        "Adamts2"       "Adamtsl1"     </span></span>
<span><span class="co">#&gt; [5] "Angpt1"        "Aqp4"</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/accessors.html">segmentIDs</a></span><span class="op">(</span><span class="va">me</span>, <span class="st">"cell"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs</span></span>
<span><span class="co">#&gt; [1] "67500" "67512" "67515" "67521" "67527"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs</span></span>
<span><span class="co">#&gt; [1] "65043" "65044" "65051" "65055" "65063" "65064" "65067" "65070" "65071"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setters">Setters<a class="anchor" aria-label="anchor" href="#setters"></a>
</h3>
<p>Main setters include <code>molecules&lt;-</code> and <code>boundaries&lt;-</code>. For example, with <code>boundaries&lt;-</code> one can add new segmentation assay information to the boundaries slot. Here we demonstrate this with the nucleus boundaries.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nucleiMEList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readBoundaries.html">readBoundaries</a></span><span class="op">(</span>dataDir <span class="op">=</span> <span class="va">repoDir</span>,</span>
<span>                            pattern <span class="op">=</span> <span class="st">"nucleus_boundaries.csv"</span>,</span>
<span>                            segmentIDCol <span class="op">=</span> <span class="st">"cell_id"</span>,</span>
<span>                            xCol <span class="op">=</span> <span class="st">"vertex_x"</span>,</span>
<span>                            yCol <span class="op">=</span> <span class="st">"vertex_y"</span>,</span>
<span>                            keepCols <span class="op">=</span> <span class="st">"essential"</span>,</span>
<span>                            boundariesAssay <span class="op">=</span> <span class="st">"nucleus"</span>,</span>
<span>                            scaleFactorVector <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>, <span class="st">"nucleus"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nucleiMEList</span></span>
<span><span class="va">me</span> <span class="co"># note the addition of the nucleus boundaries to the boundaries slot</span></span>
<span><span class="co">#&gt; class:  MoleculeExperiment </span></span>
<span><span class="co">#&gt; 2 samples: Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; @molecules contents: </span></span>
<span><span class="co">#&gt; -detected assay:</span></span>
<span><span class="co">#&gt; 178 unique features across all samples in assay "detected": 2010300C02Rik Acsbg1 Adamts2 Adamtsl1 Angpt1 Aqp4 ...</span></span>
<span><span class="co">#&gt; 869.5 molecules on average across all samples in assay "detected"</span></span>
<span><span class="co">#&gt; Location range across all samples in assay "detected": [4900,4919.98] x [6400.02,6420]</span></span></code></pre></div>
<p>The additional boundaries can be accessed, e.g. for visualisation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="../reference/accessors.html">molecules</a></span><span class="op">(</span><span class="va">me</span>, flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_location</span>, y <span class="op">=</span> <span class="va">y_location</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_polygon.html" class="external-link">geom_polygon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">segment_id</span><span class="op">)</span>,</span>
<span>                    fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                    data <span class="op">=</span> <span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>,</span>
<span>                                      assayName <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                                      flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_polygon.html" class="external-link">geom_polygon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">segment_id</span><span class="op">)</span>,</span>
<span>                    fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                    data <span class="op">=</span> <span class="fu"><a href="../reference/accessors.html">boundaries</a></span><span class="op">(</span><span class="va">me</span>,</span>
<span>                                      assayName <span class="op">=</span> <span class="st">"nucleus"</span>,</span>
<span>                                      flatten <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="MoleculeExperiment_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="from-moleculeexperiment-to-spatialexperiment">From MoleculeExperiment to SpatialExperiment<a class="anchor" aria-label="anchor" href="#from-moleculeexperiment-to-spatialexperiment"></a>
</h2>
<p>If one is interested in continuing downstream analysis at the cell-level, the MoleculeExperiment package also provides a convenience function, countMolecules(), that enables the transition from a MoleculeExperiment object to a SpatialExperiment object. With this functionality, it is possible to use already existing methods for cell-level data analysis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countMolecules.html">countMolecules</a></span><span class="op">(</span><span class="va">me</span>, boundariesAssay <span class="op">=</span> <span class="st">"nucleus"</span><span class="op">)</span></span>
<span><span class="va">spe</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 178 10 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(178): 2010300C02Rik Acsbg1 ... Trem2 Trp73</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(10): Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs.67500</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_1_outs.67512 ...</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs.65063</span></span>
<span><span class="co">#&gt;   Xenium_V1_FF_Mouse_Brain_MultiSection_2_outs.65064</span></span>
<span><span class="co">#&gt; colData names(4): sample_id x_location y_location cell_id</span></span>
<span><span class="co">#&gt; reducedDimNames(1): spatial</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_location y_location</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.4.1             MoleculeExperiment_0.99.3</span></span>
<span><span class="co">#&gt; [3] BiocStyle_2.26.0         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt;  [3] fs_1.6.1                    bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [5] rprojroot_2.0.3             GenomeInfoDb_1.34.9        </span></span>
<span><span class="co">#&gt;  [7] tools_4.2.1                 bslib_0.4.2                </span></span>
<span><span class="co">#&gt;  [9] utf8_1.2.3                  R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [11] HDF5Array_1.26.0            BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt; [13] colorspace_2.1-0            rhdf5filters_1.10.0        </span></span>
<span><span class="co">#&gt; [15] withr_2.5.0                 sp_1.5-1                   </span></span>
<span><span class="co">#&gt; [17] tidyselect_1.2.0            bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [19] compiler_4.2.1              textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [21] cli_3.6.1                   Biobase_2.58.0             </span></span>
<span><span class="co">#&gt; [23] desc_1.4.2                  DelayedArray_0.24.0        </span></span>
<span><span class="co">#&gt; [25] labeling_0.4.2              bookdown_0.33              </span></span>
<span><span class="co">#&gt; [27] sass_0.4.5                  scales_1.2.1               </span></span>
<span><span class="co">#&gt; [29] pkgdown_2.0.7               systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt; [31] stringr_1.5.0               digest_0.6.31              </span></span>
<span><span class="co">#&gt; [33] SpatialExperiment_1.8.1     rmarkdown_2.21             </span></span>
<span><span class="co">#&gt; [35] R.utils_2.12.2              XVector_0.38.0             </span></span>
<span><span class="co">#&gt; [37] pkgconfig_2.0.3             htmltools_0.5.4            </span></span>
<span><span class="co">#&gt; [39] sparseMatrixStats_1.10.0    MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt; [41] limma_3.54.1                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt; [43] highr_0.10                  rlang_1.1.0                </span></span>
<span><span class="co">#&gt; [45] DelayedMatrixStats_1.20.0   jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [47] farver_2.1.1                generics_0.1.3             </span></span>
<span><span class="co">#&gt; [49] jsonlite_1.8.4              BiocParallel_1.32.4        </span></span>
<span><span class="co">#&gt; [51] dplyr_1.1.0                 R.oo_1.25.0                </span></span>
<span><span class="co">#&gt; [53] RCurl_1.98-1.10             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [55] scuttle_1.8.0               GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">#&gt; [57] Matrix_1.5-3                Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt; [59] munsell_0.5.0               S4Vectors_0.36.1           </span></span>
<span><span class="co">#&gt; [61] Rhdf5lib_1.20.0             fansi_1.0.4                </span></span>
<span><span class="co">#&gt; [63] lifecycle_1.0.3             R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt; [65] stringi_1.7.8               yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [67] edgeR_3.40.0                SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt; [69] zlibbioc_1.44.0             rhdf5_2.42.0               </span></span>
<span><span class="co">#&gt; [71] grid_4.2.1                  parallel_4.2.1             </span></span>
<span><span class="co">#&gt; [73] dqrng_0.3.0                 lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [75] beachmat_2.14.0             magick_2.7.4               </span></span>
<span><span class="co">#&gt; [77] locfit_1.5-9.7              knitr_1.42                 </span></span>
<span><span class="co">#&gt; [79] pillar_1.9.0                GenomicRanges_1.50.2       </span></span>
<span><span class="co">#&gt; [81] rjson_0.2.21                codetools_0.2-18           </span></span>
<span><span class="co">#&gt; [83] stats4_4.2.1                glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [85] evaluate_0.20               data.table_1.14.4          </span></span>
<span><span class="co">#&gt; [87] BiocManager_1.30.20         vctrs_0.5.2                </span></span>
<span><span class="co">#&gt; [89] gtable_0.3.1                purrr_1.0.0                </span></span>
<span><span class="co">#&gt; [91] cachem_1.0.7                xfun_0.36                  </span></span>
<span><span class="co">#&gt; [93] DropletUtils_1.18.0         ragg_1.2.5                 </span></span>
<span><span class="co">#&gt; [95] SingleCellExperiment_1.20.1 tibble_3.2.0               </span></span>
<span><span class="co">#&gt; [97] memoise_2.0.1               IRanges_2.32.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bárbara Zita Peters Couto, Nicholas Robertson, Ellis Patrick, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
