<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This class enables the analysis of imaging-based ST data at the molecule
level, and standardises data across vendors. The aim of this class is to
facilitate ST data integration and comparison and, importantly, facilitate
common analytical and visualisation workflows."><title>MoleculeExperiment class: An S4 class container to store imaging-based
spatial transcriptomics data. — MoleculeExperiment-class • MoleculeExperiment</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MoleculeExperiment class: An S4 class container to store imaging-based
spatial transcriptomics data. — MoleculeExperiment-class"><meta property="og:description" content="This class enables the analysis of imaging-based ST data at the molecule
level, and standardises data across vendors. The aim of this class is to
facilitate ST data integration and comparison and, importantly, facilitate
common analytical and visualisation workflows."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MoleculeExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/MoleculeExperiment.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/SpatialUtils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>MoleculeExperiment class: An S4 class container to store imaging-based
spatial transcriptomics data.</h1>
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/SpatialUtils/blob/HEAD/R/AllClasses.R" class="external-link"><code>R/AllClasses.R</code></a>, <a href="https://github.com/SydneyBioX/SpatialUtils/blob/HEAD/R/MoleculeExperiment.R" class="external-link"><code>R/MoleculeExperiment.R</code></a></small>
      <div class="d-none name"><code>MoleculeExperiment.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This class enables the analysis of imaging-based ST data at the molecule
level, and standardises data across vendors. The aim of this class is to
facilitate ST data integration and comparison and, importantly, facilitate
common analytical and visualisation workflows.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">MoleculeExperiment</span><span class="op">(</span><span class="va">molecules</span>, boundaries <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>molecules</dt>
<dd><p>Detected transcripts information in a standardised ME list
format, as is generated by dataframeToMEList() and readMolecules() functions.</p></dd>


<dt>boundaries</dt>
<dd><p>Slot with boundary information in a standardised ME list
format, as is generated by dataframeToMEList() and readBoundaries()
functions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A MoleculeExperiment object</p>
    </div>
    <div class="section level2">
    <h2 id="slots">Slots<a class="anchor" aria-label="anchor" href="#slots"></a></h2>
    

<dl><dt><code>molecules</code></dt>
<dd><p>Slot containing information about the detected transcripts.
This slot is designed as a list of lists, where each sample contains a list
of tibbles with information for each gene. The basic information required
for this slot are the gene names of the transcripts, as well as their x and
y locations.</p></dd>


<dt><code>boundaries</code></dt>
<dd><p>Slot containing the boundaries defining each segmented cell.
The slot is designed as a list of lists, where each sample contains a list
of tibbles for each cell, consisting of the x and y coordinates of the
polygon vertices defining the cell boundary.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># creating a simple ME object from toy data</span></span></span>
<span class="r-in"><span><span class="va">molecules_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    x_coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    y_coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">boundaries_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    cell_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1"</span>, <span class="st">"cell2"</span>, <span class="st">"cell3"</span>, <span class="st">"cell4"</span>,</span></span>
<span class="r-in"><span>                    <span class="st">"cell1"</span>, <span class="st">"cell2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    vertex_x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    vertex_y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">molecules_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="dataframeToMEList.html">dataframeToMEList</a></span><span class="op">(</span><span class="va">molecules_df</span>,</span></span>
<span class="r-in"><span>                                  df_type <span class="op">=</span> <span class="st">"transcripts"</span>,</span></span>
<span class="r-in"><span>                                  assay_name <span class="op">=</span> <span class="st">"detected"</span>,</span></span>
<span class="r-in"><span>                                  sample_col <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>                                  factor_col <span class="op">=</span> <span class="st">"features"</span>,</span></span>
<span class="r-in"><span>                                  x_col <span class="op">=</span> <span class="st">"x_coords"</span>,</span></span>
<span class="r-in"><span>                                  y_col <span class="op">=</span> <span class="st">"y_coords"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">boundaries_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="dataframeToMEList.html">dataframeToMEList</a></span><span class="op">(</span><span class="va">boundaries_df</span>,</span></span>
<span class="r-in"><span>                                   df_type <span class="op">=</span> <span class="st">"boundaries"</span>,</span></span>
<span class="r-in"><span>                                   assay_name <span class="op">=</span> <span class="st">"cell"</span>,</span></span>
<span class="r-in"><span>                                   sample_col <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>                                   factor_col <span class="op">=</span> <span class="st">"cell_id"</span>,</span></span>
<span class="r-in"><span>                                   x_col <span class="op">=</span> <span class="st">"vertex_x"</span>,</span></span>
<span class="r-in"><span>                                   y_col <span class="op">=</span> <span class="st">"vertex_y"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">toy_me</span> <span class="op">&lt;-</span> <span class="fu">MoleculeExperiment</span><span class="op">(</span>molecules <span class="op">=</span> <span class="va">molecules_ls</span>,</span></span>
<span class="r-in"><span>                             boundaries <span class="op">=</span> <span class="va">boundaries_ls</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">toy_me</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class:  MoleculeExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 samples: sample1 sample2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> @molecules contents: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -detected assay:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 unique features across all samples in assay "detected": gene1 gene2 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25 molecules on average across all samples in assay "detected"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Location range across all samples in assay "detected": [0,0.97] x [0,0.98]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> @boundaries contents:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -cell:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 unique segment IDs: cell1 cell2 cell3 cell4 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Location range across all samples: [-2.22,1.73] x [-1.91,2.1]</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bárbara Zita Peters Couto, Nicholas Robertson, Ellis Patrick, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

