<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Reads in Merscope (Vizgen) molecule and boundary data from a directory,
and standardises it into a MoleculeExperiment object."><title>Read in Merscope data to an ME object — readMerscope • MoleculeExperiment</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read in Merscope data to an ME object — readMerscope"><meta property="og:description" content="Reads in Merscope (Vizgen) molecule and boundary data from a directory,
and standardises it into a MoleculeExperiment object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MoleculeExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/MoleculeExperiment.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/MoleculeExperiment/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Read in Merscope data to an ME object</h1>
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/MoleculeExperiment/blob/HEAD/R/readMerscope.R" class="external-link"><code>R/readMerscope.R</code></a></small>
      <div class="d-none name"><code>readMerscope.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads in Merscope (Vizgen) molecule and boundary data from a directory,
and standardises it into a MoleculeExperiment object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readMerscope</span><span class="op">(</span><span class="va">dataDir</span>, keepCols <span class="op">=</span> <span class="st">"essential"</span>, addBoundaries <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>dataDir</dt>
<dd><p>Character string specifying the directory with the Cosmx
output files.</p></dd>


<dt>keepCols</dt>
<dd><p>Vector of characters specifying the columns of interest from
the transcripts file. "essential" selects columns with gene names, x and y
locations. "all" will select all columns. Alternatively, specific colums
of interest can be selected by specifying them as characters in a vector.
Note that this personalised vector needs to contain the essential columns.</p></dd>


<dt>addBoundaries</dt>
<dd><p>A string with which to specify the name of the boundary
assay to be added to the me object. Can be a string, or NULL.
If NULL, a simple ME object with no boundaries will be created.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A MoleculeExperiment object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">repoDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"MoleculeExperiment"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">repoDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">repoDir</span>, <span class="st">"/vizgen_HumanOvarianCancerPatient2Slice2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">meMerscope</span> <span class="op">&lt;-</span> <span class="fu">readMerscope</span><span class="op">(</span><span class="va">repoDir</span>,</span></span>
<span class="r-in"><span>    keepCols <span class="op">=</span> <span class="st">"essential"</span>,</span></span>
<span class="r-in"><span>    addBoundaries <span class="op">=</span> <span class="st">"cell"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ⠙ 1/2 Reading transcripts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> 1/2 Reading transcripts <span style="color: #B2B2B2;">[46ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">meMerscope</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class:  MoleculeExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 samples: vizgen_HumanOvarianCancerPatient2Slice2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> @molecules contents: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -detected assay:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 486 unique features across all samples in assay <span style="font-style: italic;">detected</span>: ACKR3 ACTA2 ADAMTS4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> AKT1 AKT2 AKT3...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 15160 molecules on average across all samples in assay <span style="font-style: italic;">detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Location range across all samples in assay "detected": [10219.02,10386.83] x [8350.93,8395.3]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> @boundaries contents:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -cell:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24 unique segment IDs: 45862 45865 45872 45873 45882 45887 ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bárbara Zita Peters Couto, Nicholas Robertson, Ellis Patrick, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

